examples/traffic_histories_to_observations.py

```diff
@@ -1,130 +1,330 @@
-        interface=AgentInterface.from_type(AgentType.Laner, max_episode_steps=None),
-        agent_builder=None,
-    )
+class ObservationRecorder:
+    def __init__(
+        self,
+        scenario: str,
+        output_dir: Optional[str],
+        seed: int = 42,
+    ):
+        """Generate Observations from the perspective of one or more
+        social/history vehicles within a SMARTS scenario.
 
-    # In case we have any bubbles or additional non-history traffic
-    # in one of the scenarios, we need to add some traffic providers.
-    traffic_sims = []
-    if Scenario.any_support_sumo_traffic(scenarios):
-        sumo_traffic = SumoTrafficSimulation(headless=headless, auto_start=True)
-        traffic_sims += [sumo_traffic]
-    smarts_traffic = LocalTrafficProvider()
-    traffic_sims += [smarts_traffic]
-
-    smarts = SMARTS(
-        agent_interfaces={},
-        traffic_sims=traffic_sims,
-        envision=None if headless else Envision(),
-    )
+        Args:
+            scenario (str):
+                A path to a scenario to run.
+                Note:  the scenario should already have been built using
+                `scl scenario build ...`.
+            output_dir (str):
+                Path to the directory for the output files.
+                Will be created if necessary.
+            seed (int):
+                Seed for random number generation.  Default:  42.
+        """
+        assert scenario, "--scenario must be used to specify a scenario"
+        scenario_iter = Scenario.variations_for_all_scenario_roots([scenario], [])
+        self._scenario = next(scenario_iter)
+        assert self._scenario
+        # TAI:  also record from social vehicles?
+        assert self._scenario.traffic_history is not None
 
-    scenario_list = Scenario.get_scenario_list(scenarios)
-    scenarios_iterator = Scenario.variations_for_all_scenario_roots(scenario_list, [])
-    for scenario in scenarios_iterator:
-        obs = smarts.reset(scenario)
+        self._logger = logging.getLogger(self.__class__.__name__)
 
-        collected_data = {}
-        _record_data(smarts.elapsed_sim_time, obs, collected_data)
+        smarts_seed(seed)
+
+        self._output_dir = output_dir
+        if output_dir and not os.path.exists(output_dir):
+            os.makedirs(output_dir)
+        self._smarts = None
+        self._create_missions()
+        self._create_agent_interface()
+
+    def _create_missions(self):
+        self._missions = dict()
+        orig_missions = self._scenario.discover_missions_of_traffic_histories()
+        for v_id, mission in orig_missions.items():
+            veh_goal = self._scenario._get_vehicle_goal(v_id)
+            # TODO: get prefixed vehicle_id from TrafficHistoryProvider
+            veh_id = f"history-vehicle-{v_id}"
+            self._missions[veh_id] = replace(
+                mission, goal=PositionalGoal(veh_goal, radius=3)
+            )
+
+    def _create_agent_interface(
```

Comment:
Consider extracting the `AgentInterface` and making it as an input to the `ObservationRecorder` class.

Reply:
The `AgentInterface` is now an optional parameter to the `ObservationRecorder` class, and will default to this one if none is provided.